src/hotspot/share/opto/superword.cpp

```diff
@@ -2486,6 +2486,39 @@ VStatus VLoopBody::construct() {
   return VStatus::make_success();
 }
 
+// Returns true if the given operation can be vectorized with "truncation" where the upper bits in the integer do not
+// contribute to the result. This is true for most arithmetic operations, but false for operations such as
+// leading/trailing zero count.
+static bool can_subword_truncate(Node* in, const Type* type) {
+  if (in->is_Load() || in->is_Store() || in->is_Convert() || in->is_Phi()) {
+    return true;
+  }
+
+  int opc = in->Opcode();
+
+  // For shorts and chars, check an additional set of nodes.
+  if (type == TypeInt::SHORT || type == TypeInt::CHAR) {
+    switch (opc) {
+    case Op_ReverseBytesS:
+    case Op_ReverseBytesUS:
+      return true;
+    }
+  }
+
+  switch (opc) {
+  case Op_AddI:
+  case Op_SubI:
+  case Op_MulI:
+  case Op_AndI:
+  case Op_OrI:
+  case Op_XorI:
+    return true;
+  }
+
+  // Default to disallowing vector truncation
+  return false;
```

Comment:
I was wondering:
We could have an assert here that lists all operations that cannot be truncated?
So if a new operation is added, then we will catch that it is not handled here yet, and we can add tests, and either allow it to truncate, or add it to the list of non-truncatable operations.

Reply:
> Earlier in the function, this logic is guarded with vtn->basic_type() == T_INT so I think only integer nodes need to be added to the list. Let me know what you think!

That sounds reasonable. And that would mean we only have to add `int` operation to that assert. And if anybody ever relaxes that `vtn->basic_type() == T_INT` check, then they would immediately run into that assert. Would be nice.

Reply:
I think this is an interesting point. My concern is that this might cause unexpected assert failures to occur in rare cases, which could create a large bug tail. The behavior in the non-truncation case should not cause miscompiles so I'm not sure that a general assert is warranted. The one outlying case is that unexpected `Type` nodes can cause the later `vt = container_type(in);` line to produce an invalid result since it uses the bottom type, so I think an assert for that specifically could be useful.

Reply:
> My concern is that this might cause unexpected assert failures to occur in rare cases, which could create a large bug tail.

If it is only an assert, and you return the "conservative" answer in product, then it only affects debug. This means we don't have to be too worried about it when the assert fails. And it gives us a chance to look at that new node, and decide if truncation is ok or not. A good comment at the site of the assert would make fixing those assert bugs quite easy. It would be an assert that ensures we get better performance because we vectorize. I think this is easier than having to write IR tests for all possible nodes with all possible truncations, that would be the alternative I suppose. But with IR tests we might always miss an operation.